### ISSUE TITLE
Duplicate logs in loki

### ISSUE THREAD

**Issue by cretzel** [grafana/loki#7207]
**Describe the bug**
We see duplicate log message in grafana with loki. We are using kubernetes and promtail. We first thought this might be related to promtail config but after investigation it seems that promtail actually sends the logs only once. So we are pretty sure the issue is related to loki.

**To Reproduce**

**Expected behavior**
Log entries should only appear once.

**Environment:**
 - Infrastructure: Kubernetes
 - Deployment tool: helm
 - Loki version: 2.6.1

**Screenshots, Promtail config, or terminal output**
Loki config
```
auth_enabled: false

server:
  http_listen_port: 3100

limits_config:
  retention_period: 31d

compactor:
  working_directory: /loki/retention
  shared_store: filesystem
  compaction_interval: 10m
  retention_enabled: true
  retention_delete_delay: 2h

common:
  path_prefix: /loki
  storage:
    filesystem:
      chunks_directory: /loki/chunks
      rules_directory: /loki/rules
  replication_factor: 1
  ring:
    instance_addr: 127.0.0.1
    kvstore:
      store: inmemory

schema_config:
  configs:
    - from: 2020-10-24
      store: boltdb-shipper
      object_store: filesystem
      schema: v11
      index:
        prefix: index_
        period: 24h

ruler:
  alertmanager_url: http://localhost:9093

```

Loki HTTP API response
```
curl -G -s  "http://localhost:64747/loki/api/v1/query"   --data-urlencode 'query={job="xxx/xxx"} |= "XXX321"' | jq
{
  "status": "success",
  "data": {
    "resultType": "streams",
    "result": [
      {
        "stream": {
          "stream": "stdout",
          "app_kubernetes_io_instance": "xxx",
          "filename": "/var/log/pods/xxx/xxx/0.log",
          "level": "DEBUG",
          "log": "json",
          "namespace": "xxx",
          "pdb": "xxx-xxx",
          "pod_template_hash": "5784787cb9",
          "app_kubernetes_io_name": "xxx",
          "container_name": "xxx",
          "instance": "xxx-5784787cb9-vhjgt",
          "job": "xxx/xxx"
        },
        "values": [
          [
            "1663747467420000000",
            "[http-nio-8080-exec-8] DEBUG xxx (name=XXX321)"
          ],
          [
            "1663747467420000000",
            "[http-nio-8080-exec-8] DEBUG xxx (name=XXX321)"
          ]
        ]
      }
    ],
    "stats": {
      "summary": {
        "bytesProcessedPerSecond": 11530768,
        "linesProcessedPerSecond": 66014,
        "totalBytesProcessed": 12751,
        "totalLinesProcessed": 73,
        "execTime": 0.001105824,
        "queueTime": 5.7501e-05,
        "subqueries": 1,
        "totalEntriesReturned": 2
      },
      "querier": {
        "store": {
          "totalChunksRef": 0,
          "totalChunksDownloaded": 0,
          "chunksDownloadTime": 0,
          "chunk": {
            "headChunkBytes": 0,
            "headChunkLines": 0,
            "decompressedBytes": 0,
            "decompressedLines": 0,
            "compressedBytes": 0,
            "totalDuplicates": 0
          }
        }
      },
      "ingester": {
        "totalReached": 1,
        "totalChunksMatched": 4,
        "totalBatches": 1,
        "totalLinesSent": 2,
        "store": {
          "totalChunksRef": 0,
          "totalChunksDownloaded": 0,
          "chunksDownloadTime": 0,
          "chunk": {
            "headChunkBytes": 12751,
            "headChunkLines": 73,
            "decompressedBytes": 0,
            "decompressedLines": 0,
            "compressedBytes": 0,
            "totalDuplicates": 0
          }
        }
      }
    }
  }
}
```

**Comment by hbokh** [comment]
We see something identical on our HA setup (Loki 2.6.1 on 3 VMs (not in k8s!), spread over 3 datacenters, resulting in **triple logs**.
Part of the API query output:

```json
      {
        "stream": {
          "app": "some_app",
          "host": "some_host",
          "job": "syslog",
          "severity": "informational"
        },
        "values": [
          [
            "1664805681241832000",
            " jsonrpc.go:90: Error handling packet from 127.0.0.1:48028:  unable to fetch body for etag\n"
          ],
          [
            "1664805681241832000",
            " jsonrpc.go:90: Error handling packet from 127.0.0.1:48028:  unable to fetch body for etag\n"
          ],
          [
            "1664805681241832000",
            " jsonrpc.go:90: Error handling packet from 127.0.0.1:48028:  unable to fetch body for etag\n"
          ]
        ]
      }
```
And another example: in Grafana, the Postfix SMTP logs look like this:

![Loki-triple-errors](https://user-images.githubusercontent.com/519955/193598206-5c001a60-306c-429e-acc0-214780829053.png)

**Comment by cedsana** [comment]
I'm not sure if this is related but I also have duplicate log entries for my docker container. It's a container that runs every minute so maybe Promtail does not like that? I see 10 times each log line when I query with Loki.

**Comment by anarcher** [comment]
I have some issue with loki v2.6.1.   I didn't see this issue with 2.5.0. This issue is related to v2.6.1?

**Comment by rozhao2** [comment]
Same issue on v2.6.1 too.

some logs are duplicated 3 times, some logs are 7 times, some are 9 times.

during upgrading, promtail has no change, no restart, after upgrading done, logs duplication starts.

not all nodes' log duplicated, some nodes are good, but have no idea why.

**Comment by rucciva** [comment]
same issue, upgraded from v2.6.1 to v2.7.0 but still the same

**Comment by anarcher** [comment]
Does you have an error about `Maximum active stream limit exceeded`? After I increase active stream limit config(and downgrade to 2.5), I didn't see duplicate logs in loki (2.5).  I'm not sure if this error(`Max. active stream limit`) is related to the problem of duplicate logs.

**Comment by rucciva** [comment]
i see a lot of these line of logs from the `read` target, is it related maybe to the compaction process?

```
level=error ts=2022-12-06T06:28:37.148579567Z caller=delete_requests_client.go:211 msg="error getting delete requests from the store" err="unexpected status code: 404"
```

**Comment by dondimass** [comment]
increase Max. active stream limit help for me

**Comment by rucciva** [comment]
hi @anarcher and @dondimass , what is the config entry that you use to increase max active stream limit? is it `max_streams_per_user ` or `max_global_streams_per_user `?
thanks

**Comment by dondimass** [comment]
@rucciva i use both config entry
```
  max_streams_per_user: 100000
  max_global_streams_per_user: 100000
```

**Comment by anarcher** [comment]
I changed only the `max_global_streams_per_user`.

**Comment by StarpTech** [comment]
We experience the same on the cloud REST endpoints. If you check the source code of logcli you see they have implemented advanced deduplication + fetch logic. This is real pain from a consumer perspective. It seems it works as designed.

https://github.com/grafana/loki/blob/main/pkg/logcli/query/query.go#L142

https://github.com/grafana/loki/blob/main/pkg/logcli/query/query.go#L387

**Comment by anarcher** [comment]
Is this an issue with loki 2.7.x?  Apparently, loki 2.5 has this issue. ㅠ.ㅠ

**Comment by jjphung** [comment]
Seeing this issue in 2.7 also.

**Comment by StarpTech** [comment]
At [WunderGraph](https://wundergraph.com/) Cloud we went with a different solution. We use ClickHouse for historical data and [JetStream (Nats)](https://nats.io/) or any other queue to stream logs ordered and in real time to our users. The logs in the queue are held for only a short period of time and we fallback to ClickHouse as long-term storage. This is working very well. We will release a blog post very soon! If you're interested, please let me know, it motivates me to work on it earlier :smile:

**Comment by anarcher** [comment]
Has Loki 2.9 fixed this issue?

**Comment by cretzel** [comment]
We just went to Loki 2.9 and it fixed the duplicate logs for us.

Thanks a lot.

**Comment by cstyan** [comment]
Sounds like as of 2.9 the issue isn't being seen anymore?

**Comment by StarpTech** [comment]
As long as [this](https://github.com/grafana/loki/blob/main/pkg/logcli/query/query.go#L184) is required from a client perspective it is not fixed.

**Comment by cstyan** [comment]
Pushing data with duplicate timestamps and deduplicating at query time are separate things. If there's something that's been fixed in the web api but somehow not in logcli that's a bug that should be fixed, however the comment you're pointing to is an underlying referencing a fundamental part of how Loki's storage works. We accept log entries with the same timestamp as long as the contents are different. I don't see that changing.

**Comment by StarpTech** [comment]
> If there's something that's been fixed in the web api but somehow not in logcli that's a bug that should be fixed, however the comment you're pointing to is an underlying referencing a fundamental part of how Loki's storage works. We accept log entries with the same timestamp as long as the contents are different. I don't see that changing.

As a user, I don't care about how the loki stores the data. I want to be able to paginate through my results with the REST API. Correct me if I'm wrong but you have to implement the exact logic I referenced in the link. If you don't you have a high chance of getting duplicates.

**Comment by cstyan** [comment]
Yes, that's one of a few reasons why Loki's query API currently doesn't support pagination. If that's a feature you want we can open a separate issue for it, since that's separate from what the issue here is originally reporting.

**Comment by StarpTech** [comment]
Thanks for clarifying. I think it's really best to handle it in a separate issue.

**Comment by cstyan** [comment]
@StarpTech I've opened a new issue here https://github.com/grafana/loki/issues/11254

For now I'll close this issue. Thanks for your input :+1:
